### PR TITLE
The SSL_pretrain and DDP version of chemprop

### PR DESCRIPTION
**Pull request by jiali1025**
The SSL pretrain version of chemprop.
Here we have developed the cutomized masking and contrastive learning precedure for the directed graph of chemprop. In addition, a large mpn encoder version is created for chemprop as well, but it will only be useful when it is in SSL_pretrain mode. This is aiming at not changing other functions. Finally, the DDP version of training is enabled. This is because normally a SSL pretrain will have a huge amount of data, and to achieve a good contrastive learning performance, a huge batch size is needed. To ensure it can train fast and have huge batch size, the DDP train version is realized.

The code has passed all the tests, with two additional test, 1. test the SSL pretrain 2. test the DDP SSL pretrain.

Thanks!

### CODE REVIEW

**Comment by davidegraff** [comment]
Unfortunately, we’re no longer actively developing version 1 of chemprop, and we’re therefore unlikely to accept any PRs unless they address bugs. That said, if you would like to reimplement this feature in version 2 (which currently lives on `v2/dev` branch), we’d be willing to consider it there.

**Comment by JacksonBurns** [comment]
I'm willing to review this since it's a very nice feature to have backported.

I have approved the CI to run, and will assign myself reviewer. Please bare with me, this may take some time!

**Comment by jiali1025** [comment]
> I'm willing to review this since it's a very nice feature to have backported.
> 
> I have approved the CI to run and will assign myself reviewer. Please bare with me, this may take some time!

Hi [Jackson](https://github.com/JacksonBurns), thanks so much! I have some PowerPoints made with Kevin, which may help you understand the code more easily since there are many new codes. I will first get an updated version of that PPT first.

**Comment by JacksonBurns** [comment]
I have received the PowerPoint offline from @kevingreenman - I am going to wait to look at it until after doing code review so I can be more impartial during the review.

In the meantime, @jiali1025 please fix [this failure in the CI](https://github.com/chemprop/chemprop/actions/runs/6767624060/job/18403085449?pr=529#step:7:1150). It seems like you have a file on your machine that contains some information needed in the unit test, but it has not been added to this repository. Please add the missing file, and update the file path in the unit test.

**Comment by jiali1025** [comment]
> I have received the PowerPoint offline from @kevingreenman - I am going to wait to look at it until after doing code review so I can be more impartial during the review.
> 
> In the meantime, @jiali1025 please fix [this failure in the CI](https://github.com/chemprop/chemprop/actions/runs/6767624060/job/18403085449?pr=529#step:7:1150). It seems like you have a file on your machine that contains some information needed in the unit test, but it has not been added to this repository. Please add the missing file, and update the file path in the unit test. In addition, I have uploaded the previous ignored SSL_pretrain test data file as well. Thanks!

**Comment by jiali1025** [comment]
> For what it's worth, there's no need to port the DDP code to version as lightning [natively supports it](https://lightning.ai/docs/pytorch/stable/accelerators/gpu_intermediate.html). You'll only need to reimplement your SSL strategy in v2

Thanks, David! I will adapt it to v2 with torch lighting later

**Comment by jiali1025** [comment]
> > I have received the PowerPoint offline from @kevingreenman - I am going to wait to look at it until after doing code review so I can be more impartial during the review.
> > In the meantime, @jiali1025 please fix [this failure in the CI](https://github.com/chemprop/chemprop/actions/runs/6767624060/job/18403085449?pr=529#step:7:1150). It seems like you have a file on your machine that contains some information needed in the unit test, but it has not been added to this repository. Please add the missing file, and update the file path in the unit test.
> 
> Hi Jackson, I have fixed the bugs where I have previously assigned the TEST_DATA_DIR to my own computer location. In addition, I have uploaded the previously ignored SSL_pretrain test data file as well. Thanks!

Hi Jackson, for the failed test cases now, I think it is the SimpleNamespace object in question is likely intended to represent a configuration or a set of parameters for a test case. If the test cases are expecting a triplet_margin attribute, this implies that at some point before the test runs, the SimpleNamespace object should be populated with this attribute. The fact that it's missing indicates a disconnect between the test setup and the test execution.

**Comment by JacksonBurns** [comment]
> Thanks, David! I will adapt it to v2 with torch lighting later

That's already been done, so you will only have to move the other features to v2.

> Hi Jackson, for the failed test cases now, I think it is the SimpleNamespace object in question is likely intended to represent a configuration or a set of parameters for a test case. If the test cases are expecting a triplet_margin attribute, this implies that at some point before the test runs, the SimpleNamespace object should be populated with this attribute. The fact that it's missing indicates a disconnect between the test setup and the test execution.

I think the test failures are actually due to this new line of code:
https://github.com/chemprop/chemprop/blob/65b3308908e62bd2146b1ad9ecb9223479301e96/chemprop/train/loss_functions.py#L43

This code assumes that `triplet_margin` and `triplet_p` have always been passed by the user, which isn't the case generally and not in these units tests either. You should do something like this:
```python
"Triplet": nn.TripletMarginLoss(
   margin=getattr(args, triplet_margin, 1),
   p=getattr(args, triplet_p, 0.2),
)
```
which would use the user's input for the two, but substitute in some default values if the user doesn't provide any. Note that I haven't actually run this, but this or something like it should solve the problem.

**Comment by jiali1025** [comment]
> > Thanks, David! I will adapt it to v2 with torch lighting later
> 
> That's already been done, so you will only have to move the other features to v2.
> 
> > Hi Jackson, for the failed test cases now, I think it is the SimpleNamespace object in question is likely intended to represent a configuration or a set of parameters for a test case. If the test cases are expecting a triplet_margin attribute, this implies that at some point before the test runs, the SimpleNamespace object should be populated with this attribute. The fact that it's missing indicates a disconnect between the test setup and the test execution.
> 
> I think the test failures are actually due to this new line of code:
> 
> https://github.com/chemprop/chemprop/blob/65b3308908e62bd2146b1ad9ecb9223479301e96/chemprop/train/loss_functions.py#L43
> 
> This code assumes that `triplet_margin` and `triplet_p` have always been passed by the user, which isn't the case generally and not in these units tests either. You should do something like this:
> 
> ```python
> "Triplet": nn.TripletMarginLoss(
>    margin=getattr(args, triplet_margin, 1),
>    p=getattr(args, triplet_p, 0.2),
> )
> ```
> 
> which would use the user's input for the two, but substitute in some default values if the user doesn't provide any. Note that I haven't actually run this, but this or something like it should solve the problem.

Hi Jackson, thanks for the suggestions. I think now it is fixed

**Comment by soulios** [comment]
@jiali1025 Can I get maybe the slides so I can follow the code better? Very nice feature btw :)))

**Comment by jiali1025** [comment]
> @jiali1025 Can I get maybe the slides so I can follow the code better? Very nice feature btw :)))

Thanks!
Could u let me know your email so I can share it with you.

**Comment by JacksonBurns** [comment]
Hi @jiali1025 thank you for your patience, just letting you know I will be working on this review starting today. Excited to try this out!

File: chemprop/args.py [hunk 1]
```diff
@@ -123,6 +124,128 @@ class CommonArgs(Tap):
...
+    """
+    BD_second_percent: float = 0.3
+    """
+    For self supervised pretraining, the mask bond percentage for second view
+    """
+    SG_first_percent: float = 0.1
+    """
+    For self supervised pretraining, the subgraph deletion percentage for first view 
+    """
+    SG_second_percent: float = 0.3
```
Comment (JacksonBurns) [inline chemprop/args.py:200]: Can we group these or some of these in the same way you group the probabilities above?

File: chemprop/args.py [hunk 2]
```diff
@@ -123,6 +124,128 @@ class CommonArgs(Tap):
...
+    """
+    For self supervised pretraining, the subgraph deletion percentage for second view
+    """
+    pretrain_with_val: bool = False
+    """
+    For self supervised pretraining, whether the model is only saved with the validation score lower, not implemented well yet
```
Comment (JacksonBurns) [inline chemprop/args.py:206]: Is this still not implemented?

File: chemprop/args.py [hunk 3]
```diff
@@ -123,6 +124,128 @@ class CommonArgs(Tap):
...
+    """
+    per_depth_mpn_number: int = 1
+    """
+    The number of FFNs in each depth for the mpn
+    """
+    DDP_training: bool = False
+    """
+    Whether to use DDP training or not, necessary for large data and large model. Now it is only implemented for SSL_pretrain
+    """
+    DDP_rank: int = 0
```
Comment (JacksonBurns) [inline chemprop/args.py:236]: Why does this need to be in the CLI if it is automatically set in the background?

File: chemprop/args.py [hunk 4]
```diff
@@ -123,6 +124,128 @@ class CommonArgs(Tap):
...
+    """
+    The current rank information, it will be set dynamically for different processes
+    """
+    load_DDP: bool = False
+    """
+    Whether to load from a DDP trained model, since it will have module prefix.
+    """
+    num_DDP_gpus: int = 1
```
Comment (JacksonBurns) [inline chemprop/args.py:244]: Perhaps the default should be 2, since this is DDP?

File: chemprop/args.py [hunk 5]
```diff
@@ -123,6 +124,128 @@ class CommonArgs(Tap):
...
+    """
+    Whether to use DDP training or not, necessary for large data and large model. Now it is only implemented for SSL_pretrain
+    """
+    DDP_rank: int = 0
+    """
+    The current rank information, it will be set dynamically for different processes
+    """
+    load_DDP: bool = False
+    """
+    Whether to load from a DDP trained model, since it will have module prefix.
```
Comment (JacksonBurns) [inline chemprop/args.py:242]: If DDP-trained models have a different prefix, can't we just deduce that they are DDP models by checking for said prefix (making this arg unnecessary)?

File: chemprop/args.py [hunk 6]
```diff
@@ -339,8 +467,10 @@ class TrainArgs(CommonArgs):
     """Whether to add bias to linear layers."""
     hidden_size: int = 300
     """Dimensionality of hidden layers in MPN."""
+    mpn_encoder_middle_dim: int = 300
+    """Dimensionality of middle hidden layers in large MPN."""
     depth: int = 3
-    """Number of message passing steps."""
+    """Number of message passing steps + 1."""
```
Comment (JacksonBurns) [inline chemprop/args.py:473]: This is changing the default behavior of chemprop? This would be a significant change, if so.

(cc @kevingreenman)

File: chemprop/args.py [hunk 7]
```diff
@@ -723,6 +856,9 @@ def process_args(self) -> None:
                 self.loss_function = 'sid'
             elif self.dataset_type == 'regression':
                 self.loss_function = 'mse'
+            elif self.dataset_type == 'SSL_pretrain':
+                self.loss_function = 'MA'
+                print('the loss function of SSL_pretrain is not simple one loss function, hyperparameter tuning is available in common args, the loss_function will not be used')
```
Comment (JacksonBurns) [inline chemprop/args.py:861]: Could you rephrase what this means? I don't understand.

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
I've started by reviewing just the CLI, which I need revised before moving on to the rest of the content. Thanks again for your patience on this.

File: chemprop/args.py [hunk 8]
```diff
@@ -114,6 +114,7 @@ class CommonArgs(Tap):
     no_cache_mol: bool = False
     """
     Whether to not cache the RDKit molecule for each SMILES string to reduce memory usage (cached by default).
+    If SSL_pretrain with large dataset, better to set it to be True
```
[the remark this answers is not in the document]
Comment (jiali1025) [reply chemprop/args.py:117]: Hi Jackson, for this one, it really depends on the hardware. For my personal computer, I think once it exceeds 300K rdkit mol objects in memory it will just have the out-of-memory problem in the Mem. For SSL pre-train, normally will have more than 1M data, since no need to label the data.

File: chemprop/args.py [hunk 9]
```diff
@@ -123,6 +124,128 @@ class CommonArgs(Tap):
     """
     Path to constraints applied to atomic/bond properties prediction.
     """
+    is_pretrain: bool = False
+
```
[the remark this answers is not in the document]
Comment (jiali1025) [reply chemprop/args.py:128]: Do I need to remove this by myself and then push?

File: chemprop/args.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply chemprop/args.py:117]: Hmm ok, might be best to just always set this to True automatically for this type of training then.

File: chemprop/args.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply chemprop/args.py:128]: Yes please

**Comment by jiali1025** [comment]
@JacksonBurns Hi bro, I have fixed the conflicts and make sure it is same as master. Please let me know anything else to change

File: chemprop/args.py [hunk 10]
```diff
@@ -123,6 +124,128 @@ class CommonArgs(Tap):
...
+    """
+    Whether self supervised pretraining or not
+    """
+    not_skip_invalid_smiles: bool = False
```
[the remark this answers is not in the document]
Comment (kevingreenman) [reply chemprop/args.py:131]: I agree that this is confusing as currently written. I think the only place we deal with skipping invalid SMILES in the current code is in prediction, fingerprinting, and interpreting, all of which use `skip_invalid_smiles=False` (e.g. https://github.com/chemprop/chemprop/blob/b08063bc412f898ea2dbaf478b52ae9914604f84/chemprop/train/make_predictions.py#L62). I don't think we have an option for this in training, and the places it is used in prediction, fingerprinting, and interpreting are not exposed to the user through the CLI. To be consistent with our current training procedure, I think this option should be removed and we should assume all SMILES input by a user for training (or pretraining) are valid.

File: chemprop/args.py (hunk 6 above)
Comment (kevingreenman) [reply chemprop/args.py:465]: I don't think this is an attempt to change the default behavior, but may be a response to confusion over how this argument is used in the code [here](https://github.com/chemprop/chemprop/blame/b08063bc412f898ea2dbaf478b52ae9914604f84/chemprop/models/mpn.py#L120). Since we do message passing `for depth in range(self.depth - 1)`, I think this correction to the CLI description might be appropriate.

Comment (JacksonBurns) [reply chemprop/args.py:465]: Since range starts at 0 though, if I pass depth=3 I still get depth=3 (right?). I think changing the description to this contradicts that.

File: chemprop/args.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (JacksonBurns) [reply chemprop/args.py:131]: @jiali1025 could you change the code to work according to the way Kevin has described above?

**Comment by JacksonBurns** [comment]
> @JacksonBurns Hi bro, I have fixed the conflicts and make sure it is same as master. Please let me know anything else to change

Thanks! Please scroll through the comments I left in my review and click "Resolve Conversation" for those which you have addressed in your previous commits: https://github.com/chemprop/chemprop/pull/529#pullrequestreview-1849643309

Please also address Kevin's comment and my follow up on that comment: https://github.com/chemprop/chemprop/pull/529#discussion_r1483720197

File: chemprop/args.py (hunk 6 above)
Comment (kevingreenman) [reply chemprop/args.py:465]: If you pass `depth = 3`, it will do `for depth in range(2)`, which will be `depth=0` and `depth=1`, i.e. two message-passing iterations, right?

Comment (JacksonBurns) [reply chemprop/args.py:465]: Ah duh, thanks for typing that out for me 😅 

This seems... bad? Like we should change the `for` loop not the CLI description? Looks like we made a fencepost error in that loop when it was first written.

Comment (kevingreenman) [reply chemprop/args.py:465]: I had to think through it several times myself 😅 I agree, changing the `for` loop might make more sense, but I'm hesitant to change something significant like that right before sunsetting v1, so maybe updating the description is as far as we should go.

**Comment by jiali1025** [comment]
> > @JacksonBurns Hi bro, I have fixed the conflicts and make sure it is same as master. Please let me know anything else to change
> 
> Thanks! Please scroll through the comments I left in my review and click "Resolve Conversation" for those which you have addressed in your previous commits: [#529 (review)](https://github.com/chemprop/chemprop/pull/529#pullrequestreview-1849643309)
> 
> Please also address Kevin's comment and my follow up on that comment: [#529 (comment)](https://github.com/chemprop/chemprop/pull/529#discussion_r1483720197)

Hi bro I will do this tmr, I was sick last week, sorry for the late reply.

**Comment by jiali1025** [comment]
> > > @JacksonBurns Hi bro, I have fixed the conflicts and make sure it is same as master. Please let me know anything else to change
> > 
> > 
> > Thanks! Please scroll through the comments I left in my review and click "Resolve Conversation" for those which you have addressed in your previous commits: [#529 (review)](https://github.com/chemprop/chemprop/pull/529#pullrequestreview-1849643309)
> > Please also address Kevin's comment and my follow up on that comment: [#529 (comment)](https://github.com/chemprop/chemprop/pull/529#discussion_r1483720197)
> 
> Hi bro I will do this tmr, I was sick last week, sorry for the late reply.

@JacksonBurns Hi bro I have resolve the conversations and address Kevin's comment and your follow up comments

**Comment by JacksonBurns** [comment]
@jiali1025 thanks for you patience on this! I have been testing this using a multi-GPU machine we have locally and can confirm that it does work in DDP mode.

After consulting with @kevingreenman (and out boss), we are going to change the 'goal' for this pull request. As you can see, I have changed the base branch for this PR to a new branch called `sslddpv1`, which is just the `master` branch before #304 was merged in yesterday. Rather than resolving those merge conflicts, we are going to do a 'feature release' of SSL+DDP. This will allow users to specifically install a version of Chemprop that has these features, but it will not go into the main releases. Rather, we are going to refocus our efforts on v2.

Next steps:
 - I am going to merge your changes into this new branch, push a final commit to describe what is on the branch, set the branch to protected so that it cannot be changed, and then then release a feature version.
 - @jiali1025 please start working on implementing this feature in v2, if you are still interested. Our first release candidate should be coming out tomorrow, so you can make a branch off `v2/dev` and get going. A few notes to make that review go easier:
    - Please include user documentation in this next pull request. This has been very difficult to review because I have had to read the source code to try and figure out what this feature actually is, rather than reading a description of it and then reading the source code. You will be required to add user-facing documentation before any changes are merged into `master` anyway, so having them during the review phase makes things much easier.
    - Please do not make all the changes in a single commit. Reviewing individual commits of a hundred lines or so each is _way easier_ than review thousands of lines at once.

**Comment by jiali1025** [comment]
> Approving and merging to allow for a feature release. Looking forward to seeing this in v2!

No problem thanks! Would it be better to have the intro to SSLddv1 in the readme, so people may can find this function now as well. (As it takes time to do for v2)

**Comment by JacksonBurns** [comment]
We have added a mention of the feature release to the README in #685

**Comment by JacksonBurns** [comment]
I think if anyone is enterprising enough to try it out, they can also find the branch.
